src/worker/jobs/expiry_notification.rs

```diff
@@ -0,0 +1,135 @@
+use crate::{email::Email, models::User, worker::Environment, Emails};
+use anyhow::anyhow;
+use crates_io_worker::BackgroundJob;
+use diesel::{
+    dsl::now, Connection, ExpressionMethods, NullableExpressionMethods, PgConnection, RunQueryDsl,
+};
+use std::sync::Arc;
+
+/// The threshold in days for the expiry notification.
+const EXPIRY_THRESHOLD: i64 = 3;
+
+/// A job responsible for monitoring the status of a token.
+/// It checks if the token is about to reach its expiry date.
+/// If the token is about to expire, the job triggers a notification.
+#[derive(Default, Serialize, Deserialize, Debug)]
+pub struct CheckAboutToExpireToken;
+
+impl BackgroundJob for CheckAboutToExpireToken {
+    const JOB_NAME: &'static str = "expiry_notification";
+
+    type Context = Arc<Environment>;
+
+    #[instrument(skip(env), err)]
+    async fn run(&self, env: Self::Context) -> anyhow::Result<()> {
+        let mut conn = env.connection_pool.get()?;
+        // Check if the token is about to expire
+        // If the token is about to expire, trigger a notification.
+        check(&env.emails, &mut conn)
+    }
+}
+// Check if the token is about to expire and send a notification if it is.
+fn check(emails: &Emails, conn: &mut PgConnection) -> anyhow::Result<()> {
+    info!("Checking if tokens are about to expire");
+    let expired_tokens =
+        crate::models::token::ApiToken::find_tokens_expiring_within_days(conn, EXPIRY_THRESHOLD)?;
+    // Batch send notifications in transactions.
+    const BATCH_SIZE: usize = 100;
+    for chunk in expired_tokens.chunks(BATCH_SIZE) {
+        conn.transaction(|conn| {
+            for token in chunk {
+                // Send notification.
+                let user = User::find(conn, token.user_id)?;
+                let Some(recipient) = user.email(conn)? else {
+                    return Err(anyhow!("No address found"));
+                };
+                let email = ExpiryNotificationEmail {
+                    name: user.gh_login.clone(),
+                    token_name: token.name.clone(),
+                };
+                emails.send(&recipient, email)?;
+                // Also update the token to prevent duplicate notifications.
+                diesel::update(token)
+                    .set(crate::schema::api_tokens::expiry_notification_at.eq(now.nullable()))
+                    .execute(conn)?;
```

Comment:
After a quick look, if I'm reading correctly, this code will fail fast once an email fails to send. This behavior might also block subsequent emails that could potentially be sent successfully.

I believe it would be better to warn or collect these failed emails, skip them, and continue sending the remaining emails.